rat/variable_table.py

```diff
@@ -0,0 +1,242 @@
+from dataclasses import dataclass, field
+from enum import Enum
+import numpy
+import pandas
+import pprint
+from typing import Dict, Set, Tuple, Iterable
+
+from .exceptions import CompileError
+
+
+class VariableType(Enum):
+    DATA = 1
+    PARAM = 2
+    ASSIGNED_PARAM = 3
+
+
+@dataclass()
+class VariableRecord:
+    """
+    A record within the VariableTable
+
+    name: Name of the variable
+    variable_type: Variable type
+    base_df: The current base dataframe
+
+    subscript_rename: If the variable is renamed, save the rename here (otherwise None)
+
+    constraint_lower: Value of the lower constraint
+    constraint_upper: Value of the upper constraint
+    pad_needed: True if a padding is needed for the variable
+
+    unconstrained_vector_start_index: For parameters, the start index of the unconstrained parameter vector
+    unconstrained_vector_end_index: For parameters, the end index of the unconstrained parameter vector
+    """
+
+    name: str
+    variable_type: VariableType
+    base_df: pandas.DataFrame
+
+    subscripts_rename: Tuple[str] = field(default=None)
+
+    constraint_lower: float = field(default=float("-inf"))
+    constraint_upper: float = field(default=float("inf"))
+    pad_needed: bool = field(default=False)
+
+    unconstrained_vector_start_index: int = field(default=None, init=False)
+    unconstrained_vector_end_index: int = field(default=None, init=False)
+
+    @property
+    def subscripts(self):
+        if self.subscripts_rename is None:
+            if self.base_df is None:
+                return ()
+            else:
+                return tuple(self.base_df.columns)
+        else:
+            return self.subscripts_rename
+
+    def set_subscript_names(self, subscript_rename: Tuple[str]):
+        if self.subscripts_rename is None:
+            self.subscripts_rename = subscript_rename
+            self.base_df.columns = self.subscripts
+        else:
+            if subscript_rename != self.subscripts_rename:
+                raise AttributeError("Internal compiler error: If there are multiple renames, the rename values must match")
+
+    def add_rows_from_dataframe(self, new_rows_df: pandas.DataFrame):
+        """
+        Add rows to the base_df
+        """
+        if self.base_df is None:
+            combined_df = new_rows_df
+        else:
+            combined_df = pandas.DataFrame(numpy.concatenate([self.base_df.values, new_rows_df.values]), columns=self.base_df.columns)
+
+        self.base_df = combined_df.drop_duplicates().sort_values(list(combined_df.columns)).reset_index(drop=True)
+
+    def set_constraints(self, constraint_lower: float, constraint_upper: float):
```

Comment:
I think docstrings for the record methods should help remembering their uses in the future.

Reply:
k will add